File: ros-native/src/ros-native/src/master.cc
#include "master.h"
#include <ros/ros.h>
#include <string>

#include <json.hpp>
using json = nlohmann::json;

const char *ros_master_getHost() {
  return ros::master::getHost().c_str();
}

uint32_t ros_master_getPort() {
  return ros::master::getPort();
}

const char *ros_master_getURI() {
  return ros::master::getURI().c_str();
}

bool ros_master_check() {
  return ros::master::check();
}

const char *ros_master_getTopics() {
  ros::master::V_TopicInfo _topics;
  auto json_str = json::array();

  if (ros::master::getTopics(_topics)) {
    for (size_t i = 0; i < _topics.size(); ++i) {
      ros::master::TopicInfo ti = _topics[i];
      json_str.push_back(json{
          {"name", ti.name},
          {"datatype", ti.datatype}
      });
    }
  }

  std::string j = json_str.dump();
  char *buffer = new char[j.length()];
  std::memcpy(&buffer[0], j.c_str(), j.length());
  buffer[j.length()] = '\0';

  return &buffer[0];
}

const char *ros_master_getNodes() {
  ros::V_string _nodes;
  auto json_str = json::array();

  if (ros::master::getNodes(_nodes)) {
    for (size_t i = 0; i < _nodes.size(); ++i) {
      json_str.push_back(_nodes[i]);
    }
  }

  std::string j = json_str.dump();
  char *buffer = new char[j.length()];
  std::memcpy(&buffer[0], j.c_str(), j.length());
  buffer[j.length()] = '\0';

  return &buffer[0];
}

File: ros-sys/src/ros/master.rs
use std::ffi::CStr;
use std::ffi::CString;
use std::os::raw::c_char;

use serde_json::from_str;

use async_graphql::*;

use super::ffi;

#[derive(Clone, Debug, Default, SimpleObject, serde::Deserialize)]
pub struct TopicInfo {
  name: String,
  datatype: String,
}

#[derive(Clone, Debug, Default, SimpleObject, serde::Deserialize)]
pub struct MasterInfo {
  host: String,
  port: u32,
  uri: String,
  aliveness: bool,
  topics: Vec<TopicInfo>,
  nodes: Vec<String>,
}

pub struct Master;

impl Master {
  pub fn get_info() -> MasterInfo {
    MasterInfo {
      host: Master::get_host(),
      port: Master::get_port(),
      uri: Master::get_uri(),
      aliveness: Master::check(),
      topics: Master::get_topics(),
      nodes: Master::get_nodes(),
    }
  }

  pub fn get_host() -> String {
    unsafe {
      CStr::from_ptr(ffi::ros_master_getHost())
        .to_string_lossy()
        .into_owned()
    }
  }

  pub fn get_port() -> u32 {
    unsafe { ffi::ros_master_getPort() }
  }

  pub fn get_uri() -> String {
    unsafe {
      CStr::from_ptr(ffi::ros_master_getURI())
        .to_string_lossy()
        .into_owned()
    }
  }

  pub fn check() -> bool {
    unsafe { ffi::ros_master_check() }
  }

  pub fn get_topics() -> Vec<TopicInfo> {
    let mut topics: Vec<TopicInfo> = vec![];

    let buffer = unsafe {
      let mut ptr: *const c_char = ffi::ros_master_getTopics();
      if ptr.is_null() {
        Err("Error in ros_master_getTopics")
      } else {
        Ok(CString::from_raw(ptr as *mut c_char))
      }
    };

    match buffer {
      Ok(buff) => match buff.to_str() {
        Ok(json_str) => {
          topics = from_str::<Vec<TopicInfo>>(json_str).ok().unwrap();
        }
        Err(err) => {
          println!("{}", err);
        }
      },
      Err(err) => {
        println!("{}", err);
      }
    }

    topics
  }

  pub fn get_nodes() -> Vec<String> {
    let mut nodes: Vec<String> = vec![];

    let buffer = unsafe {
      let mut ptr: *const c_char = ffi::ros_master_getNodes();
      if ptr.is_null() {
        Err("Error in ros_master_getNodes")
      } else {
        Ok(CString::from_raw(ptr as *mut c_char))
      }
    };

    match buffer {
      Ok(buff) => match buff.to_str() {
        Ok(json_str) => {
          nodes = from_str::<Vec<String>>(json_str).ok().unwrap();
        }
        Err(err) => {
          println!("{}", err);
        }
      },
      Err(err) => {
        println!("{}", err);
      }
    }

    nodes
  }
}
